### PR TITLE
Python script to run and debug ONNX models

### PR DESCRIPTION
**Pull request by tungld**
This patch updates `debug.py` script that is used for running and debugging ONNX models.

Changing name from `debug.py` to `RunONNXModel.py`.

Adding multiple options to the script: 
- `shape_info` to set shape for dynamic inputs.
- `compile_args` to use all compiler options from `onnx-mlir` command.
- `verify` to verify the model by given reference inputs/outputs, rtol, and atol.
- `print_input`, `print_output` to print out inputs and outputs.

All supported options:
```bash
$ ONNX_MLIR_HOME=./Debug/ python ../utils/RunONNXModel.py  --help
usage: RunONNXModel.py [-h] [--compile_args COMPILE_ARGS] [--shape_info SHAPE_INFO] [--verify {onnxruntime,ref}] [--ref_folder REF_FOLDER]
                       [--rtol RTOL] [--atol ATOL]
                       model_path

positional arguments:
  model_path            Path to the ONNX model.

optional arguments:
  -h, --help            show this help message and exit
  --compile_args COMPILE_ARGS
                        Arguments passed directly to onnx-mlir command. See bin/onnx-mlir --help
  --shape_info SHAPE_INFO
                        Shape for each dynamic input, e.g. 0:1x10x20,1:7x5x3
  --verify {onnxruntime,ref}
  --ref_folder REF_FOLDER
                        Path to the folder containing reference inputs and outputs stored in protobuf. Used when --verify=ref
  --rtol RTOL           Relative tolerance for verification
  --atol ATOL           Absolute tolerance for verification
```

E.g., MNIST using random inputs:
```bash
$ ONNX_MLIR_HOME=./Debug/ python ../utils/RunONNXModel.py mnist.onnx
Temporary directory has been created at /tmp/tmpj4trc3lz
Compiling the model ...
  took  0.61645632609725  seconds.
Generating random inputs ...
  done.
Running inference ...
  took  0.0016591008752584457  seconds.
```

E.g. MNIST with reference inputs/outputs for verification, and `disable-memory-bundling`:
```bash
$ ONNX_MLIR_HOME=./Debug/ python ../utils/RunONNXModel.py mnist.onnx --compile_args="--disable-memory-bundling" --verify=ref --ref_folder=mnist/test_data_set_0
Temporary directory has been created at /tmp/tmpn4k457pp
Compiling the model ...
  took  0.5986047498881817  seconds.
Reading inputs from mnist/test_data_set_0 ...
  done.
Running inference ...
  took  0.001661023125052452  seconds.
Reading reference outputs from mnist/test_data_set_0 ...
  done.
Verifying value of Plus214_Output_0
  correct with atol=0.05, rtol=0.01.
```

In case of mismatched elements, we print out:
```
Verifying value of lstm_Y ...
Not equal to tolerance rtol=0.0005, atol=5e-08

Mismatched elements: 986 / 14000 (7.04%)
Max absolute difference: 2.4974346e-05
Max relative difference: 52.30039
 x: array([[[1.732356e-01],
        [3.278255e-07],
        [1.345366e-03],...
 y: array([[[1.732354e-01],
        [3.665352e-07],
        [1.345384e-03],...
Print out all mismatched elements ...
At (0, 0, 0) mismatch 0.1732354313135147 (actual) vs 0.17323556542396545 (reference)
At (0, 1, 0) mismatch 3.6653520396612294e-07 (actual) vs 3.2782554626464844e-07 (reference)
At (0, 2, 0) mismatch 0.0013453837018460035 (actual) vs 0.0013453662395477295 (reference)
At (0, 3, 0) mismatch 0.0001463600929128006 (actual) vs 0.00014635920524597168 (reference)
At (0, 4, 0) mismatch 0.0007426320225931704 (actual) vs 0.0007425844669342041 (reference)
At (0, 5, 0) mismatch 6.357333290907263e-07 (actual) vs 6.258487701416016e-07 (reference)
At (0, 6, 0) mismatch 0.9974887371063232 (actual) vs 0.9974887371063232 (reference)
```

### CODE REVIEW

**Comment by chentong319** [comment]
This tool will be really useful. Recently, I encountered an error in check-onnx-backend and was desperate for a tool to debug it. One issue with onnx backend test is that the output is not fully printed out. When the difference is small, I can not tell what's wrong. Can an option to show which elements are wrong with value compared?
Another question: how to generate the pb file for input? For backend test, the value of tensor is known as an bumpy tensor. Online document says that tensor flow is need to generate the pb file. But I do not have tensor flow installed.

**Comment by tungld** [comment]
@chentong319 yes, I can add that option. Thanks!

I often use this code to write a numpy array to protobuf: https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#manipulating-tensorproto-and-numpy-array

**Comment by chentong319** [comment]
> I often use this code to write a numpy array to protobuf: https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#manipulating-tensorproto-and-numpy-array

Thanks for the tip. Should it be added in the document?

Another comment: you may use `git mv` if RunONNXModel.py is modified from debug.py.

File: utils/RunONNXModel.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                start = time.perf_counter()
+                ref_session = onnxruntime.InferenceSession(temp_model_path)
+                ref_outs = ref_session.run(output_names, input_feed)
+                end = time.perf_counter()
+                print("  took ", end - start, " seconds.")
+
+            elif (verify.lower() == "ref"):
+                ref_outs = read_output_from_refs(model, ref_folder)
+            else:
+                print("Invalid verify option")
```
[the remark this answers is not in the document]
Comment (tungld) [reply utils/RunONNXModel.py:250]: Just added two options `print_input` and `print_output` for printing inputs and outputs.

**Comment by tungld** [comment]
> Thanks for the tip. Should it be added in the document?
Another comment: you may use git mv if RunONNXModel.py is modified from debug.py.

Yes, I did use `git mv` to rename the file.
I hope you have more options now: printing inputs, outputs and mismatch elements.
Thanks for your comment!

**Comment by tungld** [comment]
@chentong319 I finished all updates as requested. It is ready now. Thanks!

**Comment by tungld** [comment]
@jenkins-droid test this please

**Review by chentong319 — APPROVED** [review]
LGTM!

**Pull request merged**